lib/Sema/CSDiag.cpp

```diff
@@ -6836,25 +6836,74 @@ void FailureDiagnosis::diagnoseUnboundArchetype(ArchetypeType *archetype,
 
 /// Emit an ambiguity diagnostic about the specified expression.
 void FailureDiagnosis::diagnoseAmbiguity(Expr *E) {
+  using Archetype = std::tuple<ArchetypeType *, ConstraintLocator *, unsigned>;
 
+  llvm::SmallVector<Archetype, 2> unboundParams;
   // Check out all of the type variables lurking in the system.  If any are
   // unbound archetypes, then the problem is that it couldn't be resolved.
   for (auto tv : CS->getTypeVariables()) {
-    if (tv->getImpl().hasRepresentativeOrFixed())
+    auto &impl = tv->getImpl();
+
+    if (impl.hasRepresentativeOrFixed())
       continue;
-    
-    
+
     // If this is a conversion to a type variable used to form an archetype,
     // Then diagnose this as a generic parameter that could not be resolved.
-    auto archetype = tv->getImpl().getArchetype();
-    
+    auto archetype = impl.getArchetype();
+
     // Only diagnose archetypes that don't have a parent, i.e., ones
     // that correspond to generic parameters.
-    if (archetype && !archetype->getParent()) {
-      diagnoseUnboundArchetype(archetype, tv->getImpl().getLocator());
-      return;
+    if (!archetype || archetype->getParent())
+      continue;
+
+    // Number of constraints related to particular unbound parameter
+    // is significant indicator of the problem, because if there are
+    // no constraints associated with it, that means it can't ever be resolved,
+    // such helps to diagnose situations like: struct S<A, B> { init(_ a: A) {}}
+    // because type B would have no constraints associated with it.
+    unsigned numConstraints = 0;
+    {
+      llvm::SmallVector<Constraint *, 2> constraints;
+      CS->getConstraintGraph().gatherConstraints(tv, constraints);
+
+      for (auto constraint : constraints) {
+        // We are not interested in ConformsTo constraints because
+        // such constraints specify restrictions on the archetypes themselves.
+        if (constraint->getKind() == ConstraintKind::ConformsTo)
+          continue;
+
+        // Some of the bind constraints specify relations between
+        // parent type and it's member fields/types, we are not
+        // interested in that, since it's not related to archetype resolution.
+        if (constraint->getKind() == ConstraintKind::Bind) {
+          if (auto locator = constraint->getLocator()) {
+            auto anchor = locator->getAnchor();
+            if (anchor && isa<UnresolvedDotExpr>(anchor))
+              continue;
+          }
+        }
+
+        numConstraints++;
+      }
```

Comment:
Can you refactor this entire block of code into a separate function with a name that makes it clear what it's for?